node: kubelet: cpumgr: CPU Manager to GA

ffromani opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Changes to finalize the graduation of CPUManager to GA

#### Which issue(s) this PR fixes:
Enhancement: https://github.com/kubernetes/enhancements/issues/3570

#### Special notes for your reviewer:
As per the enhancement review, especially PRR, we need to fill gaps before to merge this PR, which is the final step to graduate to GA.
Dependencies:

- https://github.com/kubernetes/kubernetes/pull/112855

#### Does this PR introduce a user-facing change?
```release-note
Graduate Kubelet CPU Manager to GA.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
[KEP]: https://github.com/kubernetes/enhancements/pull/3570
```

ffromani committed:
node: kubelet: cpumgr: CPU Manager to GA

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/retest

fromanirh commented:
/sig node

sftim commented:
For the release note, Markdown is supported. Consider hyperlinking to https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/ or to a similar docs page.

swatisehgal commented:
/triage-accepted
/priority important-soon (as CPU Manager graduation to GA is planned for 1.26 release)

Link to the enhancement needs to be updated to https://github.com/kubernetes/enhancements/issues/3570.

swatisehgal commented:
/priority important-soon

dims commented:
Looks like all the paperwork is good for 1.26.

/approve

fromanirh commented:
I would like to have https://github.com/kubernetes/kubernetes/pull/112855 merged before this one - hence the hold

ffromani committed:
cpumanager: the FG is locked to default (ON)

hence we can remove the if() guards, the feature
is always available.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
node: cpumgr: address `golint` complains

Add docstrings and trivial fixes.

Signed-off-by: Francesco Romani <fromani@redhat.com>

fromanirh commented:
/test pull-kubernetes-e2e-gce-alpha-features

fromanirh commented:
rebased (and trivially fixed) after https://github.com/kubernetes/kubernetes/pull/112855 merged

fromanirh commented:
/hold cancel

fromanirh commented:
/retest
unrelated failure

klueska commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -310,23 +310,21 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
-		}
-		cm.topologyManager.AddHintProvider(cm.cpuManager)
+	cm.cpuManager, err = cpumanager.NewManager(
+		nodeConfig.ExperimentalCPUManagerPolicy,
+		nodeConfig.ExperimentalCPUManagerPolicyOptions,
+		nodeConfig.ExperimentalCPUManagerReconcilePeriod,
```
Should these variables still be prefixed with `Experimental` with this going to GA?

fromanirh replied:
Good point. Probably not. I'll push a separate commit to do the trivial rename

klueska commented on pkg/kubelet/cm/cpumanager/cpu_assignment.go:
```diff
@@ -27,8 +27,10 @@ import (
 	"k8s.io/kubernetes/pkg/kubelet/cm/cpuset"
 )
 
+// LoopControl control the behavior of the cpu accumulator loop logic
```
s/control/controls

fromanirh replied:
done

klueska commented on pkg/kubelet/cm/cpumanager/policy_options.go:
```diff
@@ -27,6 +27,7 @@ import (
 	"k8s.io/kubernetes/pkg/kubelet/cm/topologymanager"
 )
 
+// names of the options, as part of the user interface.
```
Capitalize `names`

fromanirh replied:
thanks, fixed

klueska commented on pkg/kubelet/cm/cpumanager/policy_options.go:
```diff
@@ -44,6 +45,8 @@ var (
 )
 
+// CheckPolicyOptionAvailable verifies if the given option can be used depending on the Feature Gate Settings.
+// returns nil on success, or a error describing the failure on error
```
s/a error/an error/

fromanirh replied:
thanks, fixed

klueska commented on pkg/kubelet/cm/cpumanager/policy_options.go:
```diff
@@ -60,6 +63,7 @@ func CheckPolicyOptionAvailable(option string) error {
 	return nil
 }
 
+// StaticPolicyOptions holds the parsed value of the policy options, ready to be consumed internally
```
period at the end

fromanirh replied:
thanks, added.

klueska commented on pkg/kubelet/cm/cpumanager/policy_options.go:
```diff
@@ -44,6 +45,8 @@ var (
 )
 
+// CheckPolicyOptionAvailable verifies if the given option can be used depending on the Feature Gate Settings.
+// returns nil on success, or a error describing the failure on error
```
period at the end

fromanirh replied:
thanks, added.

klueska commented on test/e2e_node/cpu_manager_test.go:
```diff
@@ -611,30 +594,6 @@ func runCPUManagerTests(f *framework.Framework) {
 		err = e2epod.NewPodClient(f).MatchContainerOutput(pod.Name, pod.Spec.Containers[0].Name, expAllowedCPUsListRegex)
 		framework.ExpectNoError(err, "expected log not found in container [%s] of pod [%s]",
 			pod.Spec.Containers[0].Name, pod.Name)
-
-		ginkgo.By("disable cpu manager in kubelet")
-		newCfg = configureCPUManagerInKubelet(oldCfg, &cpuManagerKubeletArguments{
-			policyName:         string(cpumanager.PolicyStatic),
-			enableCPUManager:   false,
-			reservedSystemCPUs: cpuset.CPUSet{},
-		})
-		updateKubeletConfig(f, newCfg, false)
```
Does it make sense to keep this test, but instead flip the policy to `none` instead of disabling the feature gate?

fromanirh replied:
Possibly yes, let me check.

fromanirh replied:
after some experimentation, the test would need some rethinking. We would need to cover the flow on which we change cpumanager policy across restarts, which to date has still some rough edges.
just changing the policy leads to restart failures, very likely _related_ to
```
E1102 06:43:04.969661  125395 cpu_manager.go:223] "Could not initialize checkpoint manager, please drain node and remove policy state file" err="could not restore state from checkpoint: configured policy \"none\" differs from state checkpoint policy \"static\", please drain this node and delete the CPU manager checkpoint file \"/var/lib/kubelet/cpu_manager_state\" before restarting Kubelet"
```
OTOH deleting the state file *in this test* goes against the spirit of the test itself.
All in all, I think downsizing this test and adding more tests (and possibly fixes) in separate PRs serves better the purpose.

klueska replied:
ACK

dims commented:
/approve cancel

looks like there are pending comments

dims commented:
/assign @klueska

ffromani committed:
kubelet: cpumgr: internal variable trivial rename

CPUManager is going GA, thus it makes little sense
to keep the names of the internal configuration
variables `Experimental*`.

Trivial rename only.

Signed-off-by: Francesco Romani <fromani@redhat.com>

klueska commented:
The linked enhancement in the description is incorrect, it should be:
https://github.com/kubernetes/enhancements/issues/3570

fromanirh commented:
> The linked enhancement in the description is incorrect, it should be: [kubernetes/enhancements#3570](https://github.com/kubernetes/enhancements/issues/3570)

thanks, fixed

fromanirh commented:
/hold
comment about e2e test still to be answered

fromanirh commented:
/hold cancel
comment addressed

fromanirh commented:
/retest

klueska commented:
/lgtm
/approve

klueska commented:
/cc @derekwaynecarr @dchen1107 
For final features approval

fromanirh commented:
/assign @derekwaynecarr
as feature approver for node (per OWNER_ALIASES)

dims commented:
/approve

fromanirh commented:
/retest
the failure seems unrelated

fromanirh commented:
rebased to fix conflicts

fromanirh commented:
![image](https://user-images.githubusercontent.com/25349329/199562680-aa6bb6c0-6e14-4686-a757-96d207e7ce57.png)

swatisehgal commented:
/lgtm

fromanirh commented:
/retest
